amazon/ion/core.py

```diff
@@ -411,25 +400,51 @@ def includes_second(self):
 TIMESTAMP_FRACTION_PRECISION_FIELD = 'fractional_precision'
 TIMESTAMP_FRACTIONAL_SECONDS_FIELD = 'fractional_seconds'
 MICROSECOND_PRECISION = 6
+BASE_TEN_MICROSECOND_PRECISION_EXPONENTIATION = 10 ** MICROSECOND_PRECISION
 
 
 class Timestamp(datetime):
-    """Sub-class of :class:`datetime` that supports a precision field
-
-    Notes:
-        The ``precision`` field is passed as a keyword argument of the same name.
-        The behavior of the constructor is as follows:
-            - If microsecond/fractional_precision is specified but fractional_seconds is not, a new Decimal is
-            created from the value of microsecond and truncated to fractional_precision.
-
-            - If fractional_seconds is specified but microsecond is not, microsecond is set
-            to the value of fractional_seconds scaled up to microseconds magnitude.
-
-            - A ValueError is raised when both microsecond and fractional_seconds are specified, but their values are
-            not equivalent.
-
-            - A ValueError is raised when fractional_seconds is specified, but fractional_precision is not.
-    """
+    """Sub-class of :class:`datetime` that supports a precision field; a ``fractional_precision``
+        field that specifies the  precision of the``microseconds`` field in :class:`datetime`;
+        and a ``fractional_seconds`` field that is a :class:`Decimal` specifying the fractional
+        seconds precisely.
+
+        Notes:
+            * The ``precision`` field is passed as a keyword argument of the same name.
+
+            * The ``fractional_precision`` field is passed as a keyword argument of the same name.
+              This field only relates to to the ``microseconds`` field and can be thought of
+              as the number of decimal digits that are significant.  This is an integer that
+              that is in the closed interval ``[0, 6]``.  If ``None``, ``microseconds`` must be
+              ``0`` indicating no precision below seconds.  This argument is optional and only valid
+              when ``microseconds`` is not ``None``.  If the ``microseconds`` specified has more
+              precision than this field indicates, then that is an error.
```

Comment:
This revision to the API is kind of strange, what does it mean to have it not be specified, is that invalid?  `None` and not specified should be the same (i.e. defaulted to `None`), this is also incongruent to how normal `datetime` instances work.

Put another way a normal `datetime` will not have a `fractional_precision` which is congruent to having it `None`.  Are you saying you distinguish between unspecified and `None`, that seems like a very tricky distinction to have for the API contract...